src/lib.rs

```diff
@@ -97,6 +97,23 @@ pub struct Z;
 /// Peano numbers: Increment
 pub struct S<N> ( PhantomData<N> );
 
+
+trait Pop<N> {
+    type Head;
+    type Tail;
+}
+
+impl<A, B> Pop<Z> for (A, B) {
+    type Head = A;
+    type Tail = (A, B);
+}
+
+impl<A, B: Pop<N>, N> Pop<S<N>> for (A, B) {
```

Comment:
How well are trait impls for tuples supported? We did this once and found out that it was a bit of an unstable feature - but that might have been at that time...